### ISSUE TITLE
sys_mount() can trigger set_precision() WARN_ON due to over-large buffer

### ISSUE THREAD

**Issue by leitao** [linuxppc/issues#220]
When running 'trinity'[1] syscall fuzzer on ppc64le, I am getting the following warning.

[1] https://github.com/kernelslacker/trinity
```
[  681.862099] ------------[ cut here ]------------
[  681.866793] precision 56029 too large
[  681.866888] WARNING: CPU: 1 PID: 17377 at lib/vsprintf.c:2293 set_precision+0xa0/0xc0
[  681.867099] Modules linked in: binfmt_misc ghash_generic gf128mul ecb xts ctr evdev cbc vmx_crypto virtio_balloon ip_tables x_tables autofs4 hid_generic usbhid hid ext4 crc16 mbcache jbd2 fscrypto crc32c_generic ohci_pci ehci_pci ohci_hcd ehci_hcd usbcore ibmvscsi scsi_transport_srp
[  681.867674] CPU: 1 PID: 17377 Comm: trinity-c10 Not tainted 5.0.0-rc1-00003-g9655f21d217a #960
[  681.867783] NIP:  c0000000009cd820 LR: c0000000009cd81c CTR: c0000000009d1740
[  681.867875] REGS: c00000041d7cf7a0 TRAP: 0700   Not tainted  (5.0.0-rc1-00003-g9655f21d217a)
[  681.867981] MSR:  8000000002029033 <SF,VEC,EE,ME,IR,DR,RI,LE>  CR: 28024422  XER: 20000000
[  681.868077] CFAR: c000000000106cc4 IRQMASK: 0 
[  681.868077] GPR00: c0000000009cd81c c00000041d7cfa30 c000000000e7d100 0000000000000019 
[  681.868077] GPR04: 0000000000000000 0000000000000009 000000006772616c 0000000000000019 
[  681.868077] GPR08: c000000000eb6dc0 0000000000000000 0000000000000000 c00000041d7cf75f 
[  681.868077] GPR12: 0000000000004400 c00000003ffcf480 0000000010034c30 0000000010034c20 
[  681.868077] GPR16: 0000000000000000 0000000000000000 0000000010034c40 0000000010034808 
[  681.868077] GPR20: 0000000010985da4 0000000000000433 0000000000000003 c000000000b68d76 
[  681.868077] GPR24: c00000041d7cfc88 0000000000000025 0000000000000020 0000000000000038 
[  681.868077] GPR28: c00000041d7cfbd0 c00000041d7cfc08 c000000000b68d76 c00000041d7cfad0 
[  681.868867] NIP [c0000000009cd820] set_precision+0xa0/0xc0
[  681.868930] LR [c0000000009cd81c] set_precision+0x9c/0xc0
[  681.868991] Call Trace:
[  681.869024] [c00000041d7cfa30] [c0000000009cd81c] set_precision+0x9c/0xc0 (unreliable)
[  681.869116] [c00000041d7cfab0] [c0000000009d1614] vsnprintf+0x194/0x4e0
[  681.869197] [c00000041d7cfb30] [c000000000140b2c] __request_module+0xdc/0x570
[  681.869305] [c00000041d7cfc50] [c0000000003fd0f8] get_fs_type+0xe8/0x190
[  681.869384] [c00000041d7cfcd0] [c0000000004033bc] do_mount+0x2ac/0x1040
[  681.869489] [c00000041d7cfdb0] [c000000000404668] ksys_mount+0x158/0x180
[  681.869569] [c00000041d7cfe00] [c0000000004046b0] sys_mount+0x20/0x30
[  681.869647] [c00000041d7cfe20] [c00000000000bde4] system_call+0x5c/0x70
[  681.869719] Instruction dump:
[  681.869770] 39400000 4bffffe0 7c0802a6 39200001 f8810068 3d42fff7 3c62ffd4 386368d8 
[  681.869868] 992a5bad f8010090 4b739445 60000000 <0fe00000> e8010090 e9410068 7c0803a6 
[  681.869963] ---[ end trace 9964ee192f850a9d ]---
```

**Comment by leitao** [comment]
I was able to find a testcase that hits it. It seems also that is not a ppc specific 'bug' (if we can consider this a bug).

```#define MAX     4096*10

int
main(int argc, char *argv[])
{
        int r;
        char buffer[MAX];
        memset(&buffer, 0xff, MAX);
        r = syscall(SYS_mount, "/tmp/foo", "/tmp/bar", &buffer, 0x0, NULL);                                       
}
```

**Comment by leitao** [comment]
This is a possible patch that fixes the 'problem':

http://paste.debian.net/1061284/

**Comment by mpe** [comment]
The patch seems to have disappeared?

**Comment by mpe** [comment]
This isn't a powerpc specific bug per-se, but it's caused by our PAGE_SIZE being larger than other arches. So it's probably only arm64 that's likely to also hit this.

Changing `copy_mount_string()` to only copy 4096 bytes would avoid the warning. It *shouldn't* be able to break any valid usage of `mount()` but it's still a slightly risky change.

**Comment by leitao** [comment]
I've sent this patch to fsdevel:

https://marc.info/?l=linux-fsdevel&m=154808807224059&w=2

**Comment by mpe** [comment]
Oh nice one.

You should probably mention that you found it with trinity.

**Comment by mpe** [comment]
Fixed in torvalds/linux@fbdb44013202305cd2aefb01df0a92bb55819702.

Thanks @leitao & Chandan :cookie:
